pkg/master/thirdparty_controller.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/expapi"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	thirdpartyresourceetcd "k8s.io/kubernetes/pkg/registry/thirdpartyresource/etcd"
+	"k8s.io/kubernetes/pkg/registry/thirdpartyresourcedata"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+const thirdpartyprefix = "/thirdparty/"
+
+func makeThirdPartyPath(group string) string {
+	return thirdpartyprefix + group
+}
+
+type APIInterface interface {
+	// Remove a third party resource based on the RESTful path for that resource
+	RemoveThirdPartyResource(path string) error
+	// Install a third party resource described by 'rsrc'
+	InstallThirdPartyResource(rsrc *expapi.ThirdPartyResource) error
+	// Is a particular third party resource currently installed?
+	HasThirdPartyResource(rsrc *expapi.ThirdPartyResource) (bool, error)
+	// List all currently installed third party resources
+	ListThirdPartyResources() []string
+}
+
+type ThirdPartyController struct {
+	master                     APIInterface
+	thirdPartyResourceRegistry *thirdpartyresourceetcd.REST
+}
+
+func (t *ThirdPartyController) SyncOneResource(rsrc *expapi.ThirdPartyResource) error {
+	hasResource, err := t.master.HasThirdPartyResource(rsrc)
+	if err != nil {
+		return err
+	}
+	if !hasResource {
+		return t.master.InstallThirdPartyResource(rsrc)
+	}
+	return nil
+}
+
+func (t *ThirdPartyController) SyncResources() error {
+	list, err := t.thirdPartyResourceRegistry.List(api.NewDefaultContext(), labels.Everything(), fields.Everything())
+	if err != nil {
+		return err
+	}
+	return t.syncResourceList(list)
+}
+
+func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
+	existing := sets.String{}
+	switch list := list.(type) {
+	case *expapi.ThirdPartyResourceList:
+		for ix := range list.Items {
+			item := &list.Items[ix]
+			_, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(item)
```

Comment:
I need to investigate further I guess, because this doesn't make sense to me. Can you add comments to clarify: is this looping over the *data* or the *schema* objects?